lucille-core/src/main/java/com/kmwllc/lucille/core/Document.java

```diff
@@ -51,6 +53,67 @@ public interface Document {
 
   void update(String name, UpdateMode mode, byte[]... values);
 
+  /**
+   * @throws IllegalArgumentException if object is not supported*/
+  public default void update(String name, UpdateMode mode, Object... values) throws Exception {
+    update(name, mode, (v) -> {
+        callGenericUpdater("setField", name, v);
+    }, (v) -> {
+        callGenericUpdater("setOrAdd", name, v);
+    }, values);
+  }
+
+  // TODO: put this in its own file and make it private
+  interface ThrowingConsumer<T> {
+    void accept(T t) throws Exception ;
+  }
+
+  private void update(String name, UpdateMode mode, ThrowingConsumer setter, ThrowingConsumer adder, Object... values) throws Exception {
+
+    validateFieldNames(name);
+
+    if (values.length == 0) {
+      return;
+    }
+
+    if (has(name) && mode.equals(UpdateMode.SKIP)) {
+      return;
+    }
+
+    int i = 0;
+    if (mode.equals(UpdateMode.OVERWRITE)) {
+      setter.accept(values[0]);
+      i = 1;
+    }
+    for (; i < values.length; i++) {
+      adder.accept(values[i]);
+    }
+  }
+
+  private void callGenericUpdater(String method, String name, Object value) throws Exception {
+    Class<?> clazz = null;
+
+    if (value instanceof String) {
+      clazz = String.class;
+    } else if (value instanceof Long) {
+      clazz = Long.class;
+    } else if (value instanceof Double) {
+      clazz = Double.class;
+    } else if (value instanceof Boolean) {
+      clazz = Boolean.class;
+    } else if (value instanceof Integer) {
+      clazz = Integer.class;
+    } else if (value instanceof Instant) {
+      clazz = Instant.class;
+    } else if (value instanceof byte[]) {
+      clazz = byte[].class;
+    } else {
+      throw new IllegalArgumentException(String.format("Type of Object: %s is not supported", value.toString()));
+    }
+
+    Document.class.getMethod(method, String.class, clazz).invoke(this, name, value);
```

Comment:
Java reflection has a performance cost. It may be getting faster with more recent JVMs but the cost is still there. Some pointers:

https://blogs.oracle.com/javamagazine/post/java-reflection-performance
https://www.baeldung.com/java-reflection-benefits-drawbacks#1-performance-overhead

All of the setters on Document should be considered performance-critical. They could be called in a loop. And there could be millions or billions of Documents processed in a run. So we should avoid reflection here. If there's a choice between reflection and code duplication, let's accept the duplication in favor of better performance.